northstar/src/runtime/process/mod.rs

```diff
@@ -143,32 +162,83 @@ impl Launcher {
-                    // Close writing ends of log pipes
+                    let trampoline_pid = child;
+                    // Close writing ends of log pipes (if any)
                     drop(io);
-                    // Close child console socket if any
+                    // Close child console socket (if any)
                     drop(console_fd);
                     // Close child checkpoint pipes
                     drop(checkpoint_init);
 
+                    // Receive the pid of the init process from the trampoline process
                     debug!("Waiting for init pid of {}", container);
-                    let pid = pid_channel.recv().expect("Failed to read pid") as Pid;
-
+                    let mut channel = channel.into_async_read();
+                    let pid = channel.recv::<i32>().await.expect("Failed to read pid") as Pid;
                     debug!("Created {} with pid {}", container, pid);
 
-                    // Reap the trampoline process which is (or will be) a zombie
-                    debug!("Waiting for trampoline process {} to exit", child);
-                    wait::waitpid(Some(child), None)
+                    // We're done reading the pid. The next information transferred via the
+                    // channel is the exit status of the container process.
+
+                    // Reap the trampoline process which is (or will be) a zombie otherwise
+                    debug!("Waiting for trampoline process {} to exit", trampoline_pid);
+                    wait::waitpid(Some(trampoline_pid), None)
                         .expect("Failed to wait for trampoline process");
 
-                    // Create a future that resolves upon exit of init
-                    let exit_status = waitpid(container, pid, self.tx.clone(), stop);
+                    // Start a task that waits for the exit of the init process
+                    let exit_status = {
+                        let container = container.clone();
+                        let tx = self.tx.clone();
+                        task::spawn(async move {
+                            // Wait for an event on the channel
+                            let status = match channel.recv::<ExitStatus>().await {
+                                // Init sent something
+                                Ok(exit_status) => {
+                                    debug!(
+                                        "Received exit status of {} ({}) via channel: {}",
+                                        container, pid, exit_status
+                                    );
+
+                                    // Wait for init to exit. This is needed to ensure the init process
+                                    // exited before the runtime starts to cleanup e.g remove cgroups
+                                    let pid = unistd::Pid::from_raw(pid as i32);
+                                    wait::waitpid(Some(pid), None).ok();
```

Comment:
Not sure this is even doing anything. The init process ends up being an orphan because the trampoline dies without waiting for it. This call should fail with a `ECHILD` but it is ignored either way.

Reply:
>  This call should fail with a `ECHILD` but it is ignored either way.

Good catch - have to check that. Maybe the call is too fast and init hasn't been adopted yet.

Reply:
I changed to the `ok` to a assertion on `.is_ok()`. Looks like `init` is reliably adopted by the runtime.

Reply:
well it looks like it'd be hard to be sure that such race could not ever happen in the current state.. but if you think it is fine this way then i ok too i guess ¯\_(ツ)_/¯

Reply:
well if it happens in the future we just put a `match` expression here and handle the `ECHLD` :)

Reply:
> well if it happens in the future we just put a `match` expression here and handle the `ECHLD` :)

IMHO a `ECHLD` would be a bug. Let me replace the assertion with  a match to see what `waitpid actually returned before we panic. The assertion on `is_ok` would not show that.